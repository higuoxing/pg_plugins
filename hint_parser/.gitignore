# Generated files
hint_gram.c
hint_gram.h
hint_scanner.c

# Generated subdirectories
/results/
/tmp_check/
