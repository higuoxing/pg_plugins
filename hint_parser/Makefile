MODULE_big = hint_parser

OBJS = \
	$(WIN32RES) \
	hint_gram.o \
	hint_scanner.o \
	hint_parser.o

REGRESS = hint_parser

EXTENSION = hint_parser
DATA = hint_parser--1.0.sql
PGFILEDESC = "hint_parser - Parser for query hints"

PG_CONFIG = pg_config
PGXS := $(shell $(PG_CONFIG) --pgxs)
include $(PGXS)

hint_gram.h: hint_gram.c
	touch $@

hint_gram.c: BISONFLAGS += -d

hint_gram.o hint_scanner.o: hint_gram.h
