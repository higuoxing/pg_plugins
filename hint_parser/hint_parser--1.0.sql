/* hint_parser/hint_parser--1.0.sql */

-- complain if script is sourced in psql, rather than via CREATE EXTENSION
\echo Use "CREATE EXTENSION hint_parser" to load this file. \quit

-- Entry point to parse a query string and extract a list of hints from it.
CREATE FUNCTION hint_parser(text)
RETURNS void
AS 'MODULE_PATHNAME'
LANGUAGE C;
