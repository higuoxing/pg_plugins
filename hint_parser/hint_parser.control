# hint_parser extension
comment = 'Query hint parser'
default_version = '1.0'
module_pathname = '$libdir/hint_parser'
relocatable = true
