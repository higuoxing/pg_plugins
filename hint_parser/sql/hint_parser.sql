CREATE EXTENSION hint_parser;
-- Hints with correct format
SELECT * FROM hint_parser('i(i)');
SELECT * FROM hint_parser('hint1(content1)hint2(content2)');
SELECT * FROM hint_parser('hint1( content1 ) hint2  ( content2 )');
SELECT * FROM hint_parser('  hint1 ( content1 ) h2 (c2) h3 (c3 )');
-- Valid, with newlines
SELECT * FROM hint_parser('
h1
  (
  c1
  )
h2(c2) h3(c3)
  h4(c4)
');
-- Broken hints
SELECT hint_parser('');
SELECT hint_parser('I');
SELECT hint_parser('I (AA) I2');
SELECT hint_parser('I( II ) I2( I099 ) I3 (UUKA');

DROP EXTENSION hint_parser;
