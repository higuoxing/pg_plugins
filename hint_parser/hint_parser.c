/*-------------------------------------------------------------------------
 *
 * hint_parser.c
 *		Set of functions to parse query hints from query strings.
 *
 * Copyright (c) 1996-2023, PostgreSQL Global Development Group
 *
 * IDENTIFICATION
 *		  hint_parser/hint_parser.c
 *
 *-------------------------------------------------------------------------
 */

#include "postgres.h"

#include "hint_parser.h"

#include "fmgr.h"
#include "utils/builtins.h"

PG_MODULE_MAGIC;

/*
 * This is the hint_parser function.
 */
PG_FUNCTION_INFO_V1(hint_parser);

Datum
hint_parser(PG_FUNCTION_ARGS)
{
	char	   *query = text_to_cstring(PG_GETARG_TEXT_PP(0));
	int			parse_rc;
	ListCell   *item;
	List	   *hints = NIL;

	hint_parse_result = NULL;
	hint_parse_error_msg = NULL;

	/* Parse the query */
	hint_scanner_init(query);
	parse_rc = hint_yyparse();
	hint_scanner_finish();

	if (parse_rc != 0 || hint_parse_result == NULL)
	{
		if (hint_parse_error_msg)
			elog(ERROR, "%s", hint_parse_error_msg);
		else
			elog(ERROR, "hint parser failed");
	}

	hints = hint_parse_result;

	/* Grab and print all the items parsed */
	foreach(item, hints)
	{
		HintConfigData *hint = lfirst(item);

		elog(WARNING, "Hint found: name %s contents %s",
			 hint->name, hint->contents);
	}

	PG_RETURN_VOID();
}
