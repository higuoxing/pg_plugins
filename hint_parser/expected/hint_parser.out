CREATE EXTENSION hint_parser;
-- Hints with correct format
SELECT * FROM hint_parser('i(i)');
 hint | contents 
------+----------
 i    | i
(1 row)

SELECT * FROM hint_parser('hint1(content1)hint2(content2)');
 hint  | contents 
-------+----------
 hint1 | content1
 hint2 | content2
(2 rows)

SELECT * FROM hint_parser('hint1( content1 ) hint2  ( content2 )');
 hint  | contents 
-------+----------
 hint1 | content1
 hint2 | content2
(2 rows)

SELECT * FROM hint_parser('  hint1 ( content1 ) h2 (c2) h3 (c3 )');
 hint  | contents 
-------+----------
 hint1 | content1
 h2    | c2
 h3    | c3
(3 rows)

-- Broken hints
SELECT hint_parser('');
ERROR:  syntax error at end of input
SELECT hint_parser('I');
ERROR:  syntax error at end of input
SELECT hint_parser('I (AA) I2');
ERROR:  syntax error at end of input
SELECT hint_parser('I( II ) I2( I099 ) I3 (UUKA');
ERROR:  syntax error at end of input
DROP EXTENSION hint_parser;
